packages/server-wallet/src/wallet/index.ts

```diff
@@ -497,52 +497,79 @@ export class Wallet extends EventEmitter<WalletEvent>
-    // 2. Related to one of the channels
+    // NOTE: This weird query could be avoided by adding ledgerChannelId to OpenChannel objective
+    const ledgers = channels.filter(async channel => await this.store.isLedger(channel));
+    const channelsWithPendingReqs = (
+      await Promise.all(ledgers.map(ledger => this.store.getPendingLedgerRequests(ledger)))
+    )
+      .flat()
+      .map(x => x.fundingChannelId);
 
-    const objectives = (await this.store.getObjectives(channels))
+    const objectives = (await this.store.getObjectives(channels.concat(channelsWithPendingReqs)))
       .filter(x => x !== undefined)
       .filter(o => o?.status === 'approved');
 
     while (objectives.length && !error) {
       const objective = objectives[0];
 
-      if (objective == undefined) throw new Error('Got an undefined objective '); // TODO Don't want to do this but it's a bit tricky getting type inference with Array.filter
+      // TODO: Don't want to do this but it's a bit tricky getting type inference with Array.filter
+      if (objective == undefined) throw new Error('Got an undefined objective ');
 
-      if (objective.type !== 'OpenChannel' && objective.type !== 'CloseChannel')
-        throw new Error('not implememnted');
+      let channelToLock;
 
-      const channel = objective.data.targetChannelId;
-
-      await this.store.lockApp(channel, async tx => {
-        // For the moment, we are only considering directly funded app channels.
-        // Thus, we can directly fetch the channel record, and immediately construct the protocol state from it.
-        // In the future, we can have an App model which collects all the relevant channels for an app channel,
-        // and a Ledger model which stores ledger-specific data (eg. queued requests)
-        const app = await this.store.getChannel(channel, tx);
-
-        if (!app) {
-          throw new Error('Channel not found');
-        }
+      if (objective.type === 'OpenChannel' || objective.type === 'CloseChannel')
+        channelToLock = objective.data.targetChannelId;
+      else throw new Error('crankToCompletion: unsupported objective');
 
+      await this.store.lockApp(channelToLock, async tx => {
         const setError = async (e: Error): Promise<void> => {
           error = e;
           await tx.rollback(error);
         };
+
         const markObjectiveAsDone = (): void => {
           objectives.shift();
-          channelResults.push(ChannelState.toChannelResult(app));
         };
 
+        const addChannelResult = (channel: ChannelState.ChannelState): void => {
+          channelResults.push(ChannelState.toChannelResult(channel));
+        };
+
+        let protocol;
+        let protocolState: OpenChannelProtocol.ProtocolState | CloseChannelProtocol.ProtocolState;
+
+        if (objective.type === 'OpenChannel') {
+          protocol = OpenChannelProtocol.protocol;
+          protocolState = await OpenChannelProtocol.getOpenChannelProtocolState(
```

Comment:
This is a little worrying.

The "app" is currently locked using `tx`. The call to `getOpenChannelProtocolState` might fetch that channel, or it might fetch another channel. To check that it's only fetching `channelToLock`, I would need to read its implementation.

I think that instead, either:
1. `OpenChannelProtocol.getOpenChannelProtocolState` should be a pure function of channel states; or
2. `OpenChannelProtocol.getOpenChannelProtocolState` should be used _in place of_ `lockApp`, and should accept a `criticalCode` callback?

Reply:
I agree with #2

Reply:
Let's leave it for the moment and review it when all is said and done.

Reply:
👌